File: 2015/node/day2/elves.js
const fs = require('fs');
const readline = require('readline');

// Function to read file line by line
function readFileLineByLine(filePath) {
    // Create a readable stream from the file
    const fileStream = fs.createReadStream(filePath);

    // Create a readline interface to read the file line by line
    const rl = readline.createInterface({
        input: fileStream,
        output: process.stdout,
        terminal: false
    });

    // tracking variables
    var i = 0
    var total_paper = 0
    var total_ribbon = 0

    // Listen for each line and process it
    rl.on('line', (line) => {
        // Split the line by 'x' and convert the parts to integers
        const [x1, x2, x3] = line.split('x').map(Number);

        if (isNaN(x1) || isNaN(x2) || isNaN(x3)) {
            console.error('Invalid input line:', line);
        } else {
            i += 1;
            // surface area of the box, which is 2*l*w + 2*w*h + 2*h*l
            const sides = [ x1 * x2, x1 * x3, x2 * x3 ];
            const surfaceArea = 2 * sides.reduce((a, b) => a+b, 0);
            const smallest = Math.min(...sides);
            total_paper += surfaceArea + smallest;

            // shortest distance around its sides, or the smallest perimeter of any one face
            const perimeters = [2 * x1 + 2 * x2, 2 * x1 + 2 * x3, 2 * x2 + 2 * x3];
            const smallestPerim = Math.min(...perimeters);
            // cubic feet of volume of the present
            const bow = x1 * x2 * x3;
            total_ribbon += smallestPerim + bow;
        }
    });

    // Handle the end of the file
    rl.on('close', () => {
        // console.log('Finished reading the file');
        console.log(`total paper needed for ${i} gifts: ${total_paper} ft^3`)
        console.log(`total ribbon needed is ${total_ribbon} ft`)
    });

    // Handle errors
    rl.on('error', (err) => {
        console.error('Error reading the file:', err);
    });
}

const filePath = 'input.txt';
readFileLineByLine(filePath);

File: 2015/python/day2/elves.py

if __name__ == "__main__":

    file_name = "input.txt"

    try:
        i = 0
        total_paper = 0
        total_ribbon = 0
        with open(file_name, 'r') as file:
            for line in file:
                i += 1
                dims = line.strip().split('x')
                if len(dims) == 3:
                    try:
                        x1, x2, x3 = map(int, dims)
                        sides = [x1 * x2, x1 * x3, x2 * x3]
                        # surface area of the box, which is 2*l*w + 2*w*h + 2*h*l
                        surface_area = 2 * sum(sides)
                        # the area of the smallest side
                        minimum = min(sides)
                        total_paper += surface_area + minimum

                        # shortest distance around its sides, or the smallest perimeter of any one face
                        perims = [2 * x1 + 2 * x2, 2 * x1 + 2 * x3, 2 * x2 + 2 * x3]
                        smallest_perim = min(perims)
                        # cubic feet of volume of the present
                        bow = x1 * x2 * x3
                        total_ribbon += smallest_perim + bow
                    except ValueError:
                        print(f'invalid integers at line {i}: {str(dims)}')
                else:
                    print(f'line {i} does not contain 3 numbers separated by x\'s')

        print(f'total paper needed for {i} gifts: {total_paper} ft^3')
        print(f'total ribbon needed is {total_ribbon} ft')

    except FileNotFoundError:
        print(f"The file {file_name} was not found.")
    except IOError:
        print("An error occurred while reading the file.")
    

